cmake/GetVersionFromGit.cmake

```
@@ -0,0 +1,120 @@
+set(DESCRIBE_NAME_KEY "describe-name: " CACHE STRING "Key for describe name in .git_archival.txt")
+set(COMMIT_HASH_KEY "commit: " CACHE STRING "Key for commit hash in .git_archival.txt")
+
+
+# Combine prefix and regex
+set(VERSION_REGEX_WITH_PREFIX "^${VERSION_PREFIX}${VERSION_REGEX}")
+
+# Find Git
+find_package(Git)
+
+# Attempt to retrieve version from Git
+if(EXISTS "${CMAKE_SOURCE_DIR}/.git" AND GIT_FOUND)
+    message(STATUS "Using git describe for versioning")
+
+    # Extract the version string
+    execute_process(
+        COMMAND git describe --tags --dirty
+        WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}
+        OUTPUT_VARIABLE VERSION_STRING
+        OUTPUT_STRIP_TRAILING_WHITESPACE
+        ERROR_QUIET
+    )
+
+    # If no tags are found, set version to 0 and show a warning
+    if(VERSION_STRING STREQUAL "")
+        set(VERSION_STRING "0.0.0")
+        message(WARNING 
+            "No git tags found. Version set to 0.0.0.\n"
+            "Run 'git fetch --tags' to ensure proper versioning.\n"
+            "For more information, see OpenMC developer documentation."
+        )
+    endif()
+
+    # Extract the commit hash
+    execute_process(
+        COMMAND git rev-parse HEAD
+        WORKING_DIRECTORY ${CMAKE_SOURCE_DIR}
+        OUTPUT_VARIABLE COMMIT_HASH
+        OUTPUT_STRIP_TRAILING_WHITESPACE
+    )
+else()
+    message(STATUS "Using archival file for versioning: ${ARCHIVAL_FILE}")
+    if(EXISTS "${ARCHIVAL_FILE}")
+        file(READ "${ARCHIVAL_FILE}" ARCHIVAL_CONTENT)
+
+        # Extract the describe-name line
+        string(REGEX MATCH "${DESCRIBE_NAME_KEY}([^\\n]+)" VERSION_STRING "${ARCHIVAL_CONTENT}")
+        if(VERSION_STRING MATCHES "${DESCRIBE_NAME_KEY}(.*)")
+            set(VERSION_STRING "${CMAKE_MATCH_1}")
+        else()
+            message(FATAL_ERROR "Could not extract version from ${ARCHIVAL_FILE}")
+        endif()
+
+        # Extract the commit hash
+        string(REGEX MATCH "${COMMIT_HASH_KEY}([a-f0-9]+)" COMMIT_HASH "${ARCHIVAL_CONTENT}")
+        if(COMMIT_HASH MATCHES "${COMMIT_HASH_KEY}([a-f0-9]+)")
+            set(COMMIT_HASH "${CMAKE_MATCH_1}")
+        else()
+            message(FATAL_ERROR "Could not extract commit hash from ${ARCHIVAL_FILE}")
+        endif()
+    else()
+        message(FATAL_ERROR "Neither git describe nor ${ARCHIVAL_FILE} is available for versioning.")
+    endif()
+endif()
+
+# Ensure version string format
+if(VERSION_STRING MATCHES "${VERSION_REGEX_WITH_PREFIX}")
+    set(VERSION_NO_SUFFIX "${CMAKE_MATCH_1}")
+else()
+    message(FATAL_ERROR "Invalid version format: Missing base version in ${VERSION_STRING}")
+endif()
+
+# Check for development state
+if(VERSION_STRING MATCHES "-([0-9]+)-g([0-9a-f]+)")
+    set(DEV_STATE "true")
+    set(COMMIT_COUNT "${CMAKE_MATCH_1}")
+    string(REGEX REPLACE "-([0-9]+)-g([0-9a-f]+)" "" VERSION_WITHOUT_META "${VERSION_STRING}")
```

I think you misunderstood my point originally, let's try this....

The snippet of code below used a regular expression that was NOT defined as a variable above.  It appears to be similar to, but not quite the same as the `VERSION_SUFFIX` defined above. It certainly appears to semantically be used for the purpose of finding/recording/replacing a suffix.

If the original one here is the correct regular expression, then perhaps `VERSION_SUFFIX` should be redefined to match this one?  In any case, I recommend that the "magic" regular expression here be replaced by some variable, because it is used in two places and you want to make sure it has the same value in both places.

```suggestion
if(VERSION_STRING MATCHES ${VERSION_SUFFIX})
    set(DEV_STATE "true")
    set(COMMIT_COUNT "${CMAKE_MATCH_1}")
    string(REGEX REPLACE ${VERSION_SUFFIX} "" VERSION_WITHOUT_META "${VERSION_STRING}")
```

---

Sorry for the earlier miscommunication.

To clarify, `VERSION_STRING` contains the output of `git describe`, which includes the version tag, commit count, and short hash. That’s why I used the regular expression `-([0-9]+)-g([0-9a-f]+)`, where `[0-9]+` captures the commit count and `g([0-9a-f]+)` captures the hash.

`VERSION_SUFFIX`, on the other hand, is used for a different purpose—it represents semantic suffixes like `-dev` in version tags. For example, in a tag like `v0.7.8-dev-56-gabcd1234`, `-dev` would be matched by `VERSION_SUFFIX`, while `-56-gabcd1234` comes from `git describe` and is handled by the other regex.

While the current pattern hasn’t needed modification over time, it would still be a good improvement to define it explicitly as a variable to avoid "magic" regex strings.

Thanks for pointing it out!